pkg/zrouter/zmiddlewares/topNRequests.go

```diff
@@ -0,0 +1,103 @@
+package zmiddlewares
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"github.com/zondax/golem/pkg/logger"
+	"github.com/zondax/golem/pkg/zcache"
+	"github.com/zondax/golem/pkg/zrouter/auth"
+	"net/http"
+	"strings"
+	"time"
+)
+
+const (
+	tokenDetailsTTLDefault = 45 * time.Minute
+	PathUsageByJWTKey      = "jwt_path_usage"
+	defaultTTL             = time.Hour
+)
+
+func TopRequestTokensMiddleware(zCache zcache.RemoteCache, tokenDetailsTTL, usageMetricTTL time.Duration) func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			token, err := extractBearerToken(r)
+			if err != nil {
+				logger.GetLoggerFromContext(r.Context()).Errorf("Error extracting bearer token %v", err.Error())
+			}
+
+			if token != "" {
+				details, err := getTokenDetails(r.Context(), zCache, token, tokenDetailsTTL)
+				if err != nil {
+					logger.GetLoggerFromContext(r.Context()).Errorf("Error getting token details %v", err.Error())
+				}
+
+				if details.JTI != "" {
+					if usageMetricTTL == 0 {
+						usageMetricTTL = defaultTTL
+					}
+					incrementUsageCount(r.Context(), zCache, details.JTI, r.URL.Path, usageMetricTTL)
+				}
+			}
+
+			next.ServeHTTP(w, r)
+		})
+	}
+}
+
+func extractBearerToken(r *http.Request) (string, error) {
+	authHeader := r.Header.Get(auth.Header)
+	if authHeader == "" {
+		return "", errors.New("authorization header is missing")
+	}
+
+	tokenParts := strings.Split(authHeader, " ")
+	if len(tokenParts) != 2 || tokenParts[0] != "Bearer" {
+		return "", errors.New("invalid authorization header format")
+	}
+
+	return tokenParts[1], nil
+}
+
+func getTokenDetails(ctx context.Context, zCache zcache.ZCache, token string, tokenDetailsTTL time.Duration) (auth.TokenDetails, error) {
```

Comment:
This fn could actually rely on local cache, or combined cache, instead of remote cache. What do you think?

Reply:
Maybe, but if we want to use combinedCache, we need to implement the `ZIncrBy` and `ZRevRangeWithScores` methods on localCache